<a href="https://colab.research.google.com/github/mark76jx17/BIG-DATA-project-/blob/main/BIG_DATA_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Download Points of Interest (POIs) for Health, Education, Food, Security, Public Services, and Sports in Pavia and Cagliari using `osmnx`, calculate H3 indices for these locations to aggregate service counts per hexagonal cell, and create an interactive KeplerGL map to visualize the density of basic services in both cities.

In [7]:
from google.colab import output
output.enable_custom_widget_manager()

## Install Libraries
Install necessary Python libraries including `osmnx`, `h3`, `geopandas`, and `keplergl` to handle OpenStreetMap data, spatial indexing, and visualization.


In [ ]:
!pip install osmnx h3 geopandas keplergl

## Download OSM Data


Download Points of Interest (POIs) for Pavia and Cagliari using `osmnx`.


In [3]:
import osmnx as ox
import pandas as pd
import geopandas as gpd
import warnings

# Suppress OSMnx UserWarnings regarding query area size to prevent stderr output
warnings.filterwarnings("ignore", category=UserWarning, module="osmnx")

# Define locations
locations = ['Pavia, Italy', 'Cagliari, Italy']

# Define tags
tags = {
    'amenity': [
        'hospital', 'clinic', 'doctors', 'pharmacy',
        'school', 'university', 'kindergarten', 'college', 'library',
        'restaurant', 'cafe', 'fast_food', 'bar', 'pub',
        'police', 'fire_station', 'post_office', 'townhall', 'courthouse'
    ],
    'leisure': [
        'sports_centre', 'pitch', 'stadium', 'swimming_pool'
    ]
}

# Download POIs
gdf_pois = ox.features.features_from_place(locations, tags)

# Verify the data
print(f"Shape of GeoDataFrame: {gdf_pois.shape}")
display(gdf_pois.head())

Shape of GeoDataFrame: (6848, 243)


geometry      amenity created_by addr:city  \
element id                                                                      
node    207112603   POINT (9.1538 45.18236)     pharmacy        NaN     Pavia   
        207145209  POINT (9.15732 45.18591)  post_office        NaN       NaN   
        207559055   POINT (9.1635 45.18536)   university        NaN       NaN   
        230030507  POINT (9.17271 44.95407)   restaurant       JOSM       NaN   
        230120869  POINT (9.16048 44.90298)         cafe        NaN       NaN   

                  addr:housenumber addr:postcode         addr:street  \
element id                                                             
node    207112603               12         27100  Corso Strada Nuova   
        207145209              NaN           NaN                 NaN   
        207559055              NaN           NaN                 NaN   
        230030507              NaN           NaN                 NaN   
        230120869              NaN           NaN                 NaN   

                                                        name operator  \
element id                                                              
node    207112603                           Farmacia Tonello      NaN   
        207145209                                        NaN      NaN   
        207559055  Università di Pavia - Facoltà di Botanica      NaN   
        230030507                        Pizzeria L'Orchidea      NaN   
        230120869                 Circolo ANSPI Sant'Eusebio      NaN   

                  operator:wikidata  ... building:part  \
element id                           ...                 
node    207112603               NaN  ...           NaN   
        207145209               NaN  ...           NaN   
        207559055               NaN  ...           NaN   
        230030507               NaN  ...           NaN   
        230120869               NaN  ...           NaN   

                  it:lombardia:ctrn:cascina:CS_ID note:payment:meal_voucher  \
element id                                                                    
node    207112603                             NaN                       NaN   
        207145209                             NaN                       NaN   
        207559055                             NaN                       NaN   
        230030507                             NaN                       NaN   
        230120869                             NaN                       NaN   

                  public motorcycle:rental motorcycle:training  \
element id                                                       
node    207112603    NaN               NaN                 NaN   
        207145209    NaN               NaN                 NaN   
        207559055    NaN               NaN                 NaN   
        230030507    NaN               NaN                 NaN   
        230120869    NaN               NaN                 NaN   

                  motorcycle:type contact:telegram climbing:toprope  \
element id                                                            
node    207112603             NaN              NaN              NaN   
        207145209             NaN              NaN              NaN   
        207559055             NaN              NaN              NaN   
        230030507             NaN              NaN              NaN   
        230120869             NaN              NaN              NaN   

                  residential  
element id                     
node    207112603         NaN  
        207145209         NaN  
        207559055         NaN  
        230030507         NaN  
        230120869         NaN  

[5 rows x 243 columns]

## Process Data and Calculate H3 Indices
Convert POI geometries to centroids and calculate the H3 index for each location at resolution 9.


In [4]:
import h3
import warnings

# Suppress GeoPandas UserWarning about centroids on geographic CRS
warnings.filterwarnings("ignore", message="Geometry is in a geographic CRS")

# Create a copy to avoid modifying the original dataframe
gdf_h3 = gdf_pois.copy()

# Convert geometries to centroids
gdf_h3['geometry'] = gdf_h3.geometry.centroid

# Extract latitude and longitude
gdf_h3['lat'] = gdf_h3.geometry.y
gdf_h3['lng'] = gdf_h3.geometry.x

# Calculate H3 index at resolution 9
resolution = 9
gdf_h3['h3_index'] = gdf_h3.apply(lambda row: h3.latlng_to_cell(row['lat'], row['lng'], resolution), axis=1)

# Verify the calculation
display(gdf_h3[['geometry', 'lat', 'lng', 'h3_index']].head())

geometry        lat       lng  \
element id                                                         
node    207112603   POINT (9.1538 45.18236)  45.182364  9.153800   
        207145209  POINT (9.15732 45.18591)  45.185905  9.157324   
        207559055   POINT (9.1635 45.18536)  45.185359  9.163501   
        230030507  POINT (9.17271 44.95407)  44.954066  9.172706   
        230120869  POINT (9.16048 44.90298)  44.902976  9.160479   

                          h3_index  
element id                          
node    207112603  891f994133bffff  
        207145209  891f9941333ffff  
        207559055  891f9941337ffff  
        230030507  891f9bb1547ffff  
        230120869  891f9bbac1bffff

## Aggregate Service Counts

Group the POI data by their H3 cell index to calculate the total number of services available in each hexagonal unit.


In [5]:
# Group by H3 index and count services
df_counts = gdf_h3.groupby('h3_index').size().reset_index(name='service_count')

# Verify the aggregated data
print(f"Number of unique H3 cells: {df_counts.shape[0]}")
display(df_counts.head())

Number of unique H3 cells: 2673


h3_index  service_count
0  891e9299003ffff              3
1  891e9299007ffff             11
2  891e929900bffff              6
3  891e929900fffff              4
4  891e9299013ffff              2

## Generate Interactive Map

Create an interactive KeplerGL map to visualize the density of basic services per H3 cell for both cities.


In [6]:
from keplergl import KeplerGl

# Initialize KeplerGl map
map_1 = KeplerGl(height=600)

# Add data to the map
map_1.add_data(data=df_counts, name='Service Density')

# Save the map to an HTML file
map_1.save_to_html(file_name='services_map.html')

# Display the map
map_1

User Guide: https://docs.kepler.gl/docs/keplergl-jupyter
Map saved to services_map.html!


KeplerGl(data={'Service Density': {'index': [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18,…

In [10]:
from google.colab import files
files.download('services_map.html')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>


### Data Analysis Key Findings

*   **Data Retrieval**: downloaded Points of Interest (POIs) for categories including Health, Education, Food, Security, Public Services, and Sports in Pavia and Cagliari. The resulting dataset contained **6,848** distinct locations.
*   **Spatial Indexing**: Converted POI geometries to centroids and calculated H3 spatial indices at **resolution 9** for every location.
*   **Aggregation**: Aggregated the service data based on spatial indices, resulting in **2,673** unique hexagonal cells containing at least one service.
*   **Visualization Output**: Generated an interactive KeplerGL map (`services_map.html`) visualizing the density of basic services ('service_count') across the hexagonal grid.

### Insights or Next Steps

*   **Urban Density Analysis**: The generated map provides a visual heatmap of service availability, allowing for the immediate identification of "15-minute city" potential by highlighting areas with high concentrations of essential services versus underserved zones in both Pavia and Cagliari.
*   **Granular Filtering**: A recommended next step is to separate the aggregation by specific tag categories (e.g., separating "Health" from "Leisure") to analyze the distribution of specific infrastructure types rather than just total service density.
